File: my-folder/1250-longest-common-subsequence/solution.cpp
class Solution {
public:

    int solve(int i,int j,string &s1,string &s2,vector<vector<int>> &dp){
        if(i < 0 || j < 0)return 0;

        if(i == 0 && j == 0){
            return s1[i] == s2[j];
        }
        if(dp[i][j] != -1)return dp[i][j];
        int one = 0;

        if(s1[i] == s2[j]){
            one = 1 + solve(i - 1,j - 1,s1,s2,dp);
        }
        int two = solve(i-1,j,s1,s2,dp);
        int three = solve(i,j - 1,s1,s2,dp);

        return dp[i][j] = max({one ,two,three});
    }

    int longestCommonSubsequence(string s1, string s2) {
        int n1 = s1.size(),n2 = s2.size();
        vector<vector<int>> dp(n1 + 1,vector<int> (n2 + 1));

        for(int i = 0;i<n1;i++)dp[i][0] = 0;
        for(int i = 0;i<n2;i++)dp[0][i] = 0;

            for(int i= 1;i<=n1;i++){
                for(int j = 1;j<=n2;j++){
                    int one = 0;
                    if(s1[i - 1] == s2[j - 1]) one =  1  + dp[i - 1][j - 1];
                    int two = max(dp[i - 1][j],dp[i][j - 1]);

                    dp[i][j] = max(one,two);
                }

            }

            return dp[n1][n2];
            
        }
    };

File: my-folder/1250-longest-common-subsequence/solution.java
class Solution {

    int solve(int i,int j,int dp[][],String s1,String s2){
        
        if(i < 0 || j < 0)return 0;
        if(dp[i][j] != -1)return dp[i][j];
        int one =0;

        if(s1.charAt(i) == s2.charAt(j)){
            one =  1 + solve(i - 1,j - 1,dp,s1,s2); 
        }

        return dp[i][j] = Math.max(one,Math.max(solve(i - 1,j,dp,s1,s2),solve(i,j - 1,dp,s1,s2)));


    }

    public int longestCommonSubsequence(String s1, String s2) {
        int n = s1.length();
        int m = s2.length() ;
        int [][] dp = new int[n + 1][m+ 1];
        for(int i= 0;i<n;i++){
            for(int j = 0;j<m;j++){
                dp[i][j] = -1;
            }
        }
        return solve(n - 1,m - 1,dp,s1,s2);
        

    }
}
